File: cpr-fantasy-football/lib/payment-processor.ts
import { fetchCSV } from './data-fetcher';
import { CSV_URLS } from './constants';

export interface PlayerPaymentDetail {
  name: string;
  matchFees: number;
  fines: number;
  totalOwed: number;
  paid: number;
  balance: number;
  matchCount: number;
  fineDetails: { date: string; amount: number; description: string }[];
  paymentDetails: { date: string; amount: number }[];
}

// Parse currency string (e.g., "£12.00" -> 12.00)
function parseCurrency(value: string): number {
  if (!value) return 0;
  const cleaned = String(value).replace(/[£,]/g, '').trim();
  const parsed = parseFloat(cleaned);
  return isNaN(parsed) ? 0 : parsed;
}

// Normalize player names for matching
function normalizePlayerName(name: string): string {
  if (!name) return '';
  return name.toLowerCase().trim();
}

// Extract player name from bank statement description
function extractPlayerName(description: string): string {
  if (!description) return '';

  // Try to extract name from various formats
  // "BRUCE A J W/JUN07 , ANDREW BRUCE , ..."
  // "J WALKER , FOOTBALL , ..."
  // "G LAKE , FEES , ..."

  const parts = description.split(',');
  if (parts.length > 0) {
    const firstPart = parts[0].trim();
    // Remove dates and extra info
    const cleaned = firstPart.replace(/\/[A-Z]+\d+/gi, '').trim();
    return cleaned;
  }

  return description.trim();
}

// Map common name variations
function mapPlayerName(name: string): string {
  const normalized = normalizePlayerName(name);

  // Add common variations here if needed
  const nameMap: { [key: string]: string } = {
    'j young': 'j. young',
    'joshua young': 'j. young',
    'g lake': 'g. lake',
    'george lake': 'g. lake',
    // Add more mappings as needed
  };

  return nameMap[normalized] || name;
}

export async function getPlayerPayments(): Promise<PlayerPaymentDetail[]> {
  try {
    // Fetch all data sources
    const [matchData, finesData, bankData] = await Promise.all([
      fetchCSV<any>(CSV_URLS.MATCH_DETAILS),
      fetchCSV<any>(CSV_URLS.FINES),
      fetchCSV<any>(CSV_URLS.BANK_STATEMENT),
    ]);

    const playerMap = new Map<string, PlayerPaymentDetail>();

    // Process match fees (skip first 3 rows)
    const matchRows = matchData.slice(3);
    matchRows.forEach((row: any) => {
      const playerName = String(row._5 || '').trim();
      const fee = parseCurrency(row._2);
      const date = String(row._1 || '').trim();

      if (!playerName || fee === 0) return;

      const normalized = normalizePlayerName(playerName);
      if (!playerMap.has(normalized)) {
        playerMap.set(normalized, {
          name: playerName,
          matchFees: 0,
          fines: 0,
          totalOwed: 0,
          paid: 0,
          balance: 0,
          matchCount: 0,
          fineDetails: [],
          paymentDetails: [],
        });
      }

      const player = playerMap.get(normalized)!;
      player.matchFees += fee;
      player.matchCount += 1;
    });

    // Process fines (skip first 4 rows)
    const fineRows = finesData.slice(4);
    fineRows.forEach((row: any) => {
      const playerName = String(row._4 || '').trim();
      const fine = parseCurrency(row._2);
      const date = String(row._1 || '').trim();
      const description = String(row._3 || '').trim();

      if (!playerName || fine === 0) return;

      const normalized = normalizePlayerName(playerName);
      if (!playerMap.has(normalized)) {
        playerMap.set(normalized, {
          name: playerName,
          matchFees: 0,
          fines: 0,
          totalOwed: 0,
          paid: 0,
          balance: 0,
          matchCount: 0,
          fineDetails: [],
          paymentDetails: [],
        });
      }

      const player = playerMap.get(normalized)!;
      player.fines += fine;
      player.fineDetails.push({ date, amount: fine, description });
    });

    // Process bank payments (skip first row if header)
    bankData.forEach((row: any) => {
      const description = String(row.Description || '').trim();
      const payment = parseCurrency(row.In);
      const date = String(row.Date || '').trim();

      if (!description || payment === 0) return;

      // Try to match player name from description
      const extractedName = extractPlayerName(description);

      // Try to find matching player in our map
      let matchedPlayer: PlayerPaymentDetail | undefined;
      let matchedKey = '';

      for (const [key, player] of playerMap.entries()) {
        const playerNameLower = normalizePlayerName(player.name);
        const extractedNameLower = normalizePlayerName(extractedName);

        // Check if extracted name contains player name or vice versa
        if (extractedNameLower.includes(playerNameLower) ||
            playerNameLower.includes(extractedNameLower) ||
            extractedName.toLowerCase().includes(key)) {
          matchedPlayer = player;
          matchedKey = key;
          break;
        }
      }

      if (matchedPlayer) {
        matchedPlayer.paid += payment;
        matchedPlayer.paymentDetails.push({ date, amount: payment });
      }
    });

    // Calculate totals and balances
    for (const player of playerMap.values()) {
      player.totalOwed = player.matchFees + player.fines;
      player.balance = player.totalOwed - player.paid;
    }

    // Convert to array and sort by balance (highest debt first)
    const players = Array.from(playerMap.values());
    players.sort((a, b) => b.balance - a.balance);

    return players;
  } catch (error) {
    console.error('Error processing player payments:', error);
    throw error;
  }
}

File: conda-fantasy-football/src/payment_processor.py
"""Payment data processing."""

import pandas as pd
from datetime import datetime
from typing import Dict, List
from .data_fetcher import fetch_csv, parse_number
from .config import CSV_URLS, MATCH_COLUMNS, BANK_COLUMNS, SEASON_CONFIG


def normalize_player_name(name: str) -> str:
    """Normalize player names for matching across different data sources."""
    if not name or pd.isna(name):
        return ""
    return str(name).lower().strip()


def get_player_payments() -> pd.DataFrame:
    """
    Fetch and process player payment information.
    Combines data from player data, match fees, and bank statement.

    Returns:
        DataFrame with player payment details sorted by balance (highest debt first)
    """
    # Fetch all data sources
    player_data = fetch_csv(CSV_URLS["PLAYER_DATA"])
    match_data = fetch_csv(CSV_URLS["MATCH_DETAILS"])
    bank_data = fetch_csv(CSV_URLS["BANK_STATEMENT"], skip_rows=0)  # No headers
    fines_data = fetch_csv(CSV_URLS["FINES"], skip_rows=0)  # No headers

    player_map: Dict[str, Dict] = {}

    # Process player data (has Fees, Payments, Due columns)
    for _, row in player_data.iterrows():
        player_name = str(row.get("Player", "")).strip()
        fees = parse_number(row.get("Fees", 0))
        payments = parse_number(row.get("Payments", 0))
        due = parse_number(row.get("Due", 0))
        appearance = int(parse_number(row.get("Appearance", 0)))

        if not player_name:
            continue

        normalized = normalize_player_name(player_name)
        player_map[normalized] = {
            "name": player_name,
            "match_fees": fees,
            "season_fees": 0,  # Will be calculated later
            "fines": 0,  # Will be calculated from fines CSV
            "total_owed": fees,
            "paid": payments,
            "balance": due,
            "match_count": appearance,
            "match_details": [],
            "payment_details": [],
            "fine_details": [],
        }

    # Process match details to get match fees per game (skip first 3 rows)
    match_rows = match_data.iloc[3:]
    for _, row in match_rows.iterrows():
        player_name = str(row.get(MATCH_COLUMNS["PLAYER"], "")).strip()
        fee = parse_number(row.get(MATCH_COLUMNS["FEE"], 0))
        date = str(row.get(MATCH_COLUMNS["DATE"], "")).strip()
        game = str(row.get(MATCH_COLUMNS["GAME"], "")).strip()

        if not player_name or not date:
            continue

        normalized = normalize_player_name(player_name)
        player = player_map.get(normalized)

        if player and fee > 0:
            player["match_details"].append({"date": date, "fee": fee, "game": game})

    # Process bank payments
    # Bank statement CSV has no headers, we need to skip the first row if it's a header
    # Let's check if first row looks like a header
    if not bank_data.empty:
        first_row = bank_data.iloc[0]
        # If first column contains "Date" or similar, skip it
        if str(first_row.iloc[0]).strip().lower() in ["date", "Date"]:
            bank_data = bank_data.iloc[1:]

    for _, row in bank_data.iterrows():
        # Access by position using column indices
        try:
            date = str(row.iloc[BANK_COLUMNS["DATE"]]).strip() if len(row) > BANK_COLUMNS["DATE"] else ""
            description = str(row.iloc[BANK_COLUMNS["DESCRIPTION"]]).strip() if len(row) > BANK_COLUMNS["DESCRIPTION"] else ""
            payment = parse_number(row.iloc[BANK_COLUMNS["CREDIT"]] if len(row) > BANK_COLUMNS["CREDIT"] else 0)
            player_name_from_bank = str(row.iloc[BANK_COLUMNS["PLAYER"]]).strip() if len(row) > BANK_COLUMNS["PLAYER"] else ""

            # Skip if no player name or payment
            if not player_name_from_bank or payment == 0 or not date:
                continue

            # Parse date and filter for dates from payment start date onwards
            date_parts = date.split("/")
            if len(date_parts) != 3:
                continue

            day, month, year = map(int, date_parts)
            row_date = datetime(year, month, day)

            # Only process dates from the configured payment start date onwards
            if row_date < SEASON_CONFIG["PAYMENT_START_DATE"]:
                continue

            # Match player
            normalized = normalize_player_name(player_name_from_bank)
            player = player_map.get(normalized)

            if player:
                player["payment_details"].append(
                    {"date": date, "amount": payment, "description": description}
                )
        except (IndexError, ValueError):
            continue

    # Process fines data (skip first 4 rows)
    if not fines_data.empty:
        fines_rows = fines_data.iloc[4:]
        for _, row in fines_rows.iterrows():
            try:
                # Column indices: 0 is empty, 1 is Date, 2 is Fines, 3 is Description, 4 is Player
                player_name = str(row.iloc[4]).strip() if len(row) > 4 else ""
                fine = parse_number(row.iloc[2] if len(row) > 2 else 0)
                date = str(row.iloc[1]).strip() if len(row) > 1 else ""
                description = str(row.iloc[3]).strip() if len(row) > 3 else ""

                if not player_name or not date or fine <= 0:
                    continue

                normalized = normalize_player_name(player_name)
                player = player_map.get(normalized)

                if player:
                    player["fines"] += fine
                    player["fine_details"].append(
                        {"date": date, "amount": fine, "description": description}
                    )
            except (IndexError, ValueError):
                continue

    # Calculate season fees and update totals
    for player in player_map.values():
        # Apply season fee if player has played more than threshold number of games
        player["season_fees"] = (
            SEASON_CONFIG["SEASON_FEE"]
            if player["match_count"] > SEASON_CONFIG["SEASON_FEE_THRESHOLD"]
            else 0
        )

        # Update totalOwed to include season fees
        player["total_owed"] = player["match_fees"] + player["season_fees"]

        # Recalculate balance as Total Owed - Total Paid
        player["balance"] = player["total_owed"] - player["paid"]

        # Sort details by date
        for detail_key in ["match_details", "payment_details", "fine_details"]:
            player[detail_key].sort(key=lambda x: parse_date_for_sorting(x["date"]))

    # Convert to DataFrame
    players_list = [
        p
        for p in player_map.values()
        if p["match_count"] > 0 or p["fines"] > 0
    ]

    if not players_list:
        return pd.DataFrame()

    df = pd.DataFrame(players_list)

    # Sort by balance (highest debt first)
    df = df.sort_values("balance", ascending=False).reset_index(drop=True)

    return df


def parse_date_for_sorting(date_str: str) -> tuple:
    """Parse DD/MM/YYYY format for sorting."""
    try:
        parts = date_str.split("/")
        if len(parts) == 3:
            day, month, year = map(int, parts)
            return (year, month, day)
    except:
        pass
    return (0, 0, 0)
